typed-protocols/src/Network/TypedProtocol/Driver.hs

```diff
@@ -125,16 +125,29 @@ runPipelinedPeer
   -> Codec ps failure m bytes
   -> Channel m bytes
   -> PeerPipelined ps pr st m a
-  -> m a
+  -> m (Either failure a)
 runPipelinedPeer maxOutstanding codec channel (PeerPipelined peer) = do
-    receiveQueue <- atomically $ newTBQueue maxOutstanding
-    collectQueue <- atomically $ newTBQueue maxOutstanding
-    ((), x) <- runPipelinedPeerReceiverQueue receiveQueue collectQueue
-                                             codec channel
-                 `concurrently`
-               runPipelinedPeerSender        receiveQueue collectQueue
-                                             codec channel peer
-    return x
+    trailingBytes <- atomically $ newTMVar Nothing
+    receiveQueue  <- atomically $ newTBQueue maxOutstanding
+    collectQueue  <- atomically $ newTBQueue maxOutstanding
+    withAsync
+      (runPipelinedPeerReceiverQueue
+        receiveQueue trailingBytes collectQueue
+        codec channel)
+      $ \queueAsync -> do
+          -- TODO:
+          -- We'd like to link the queue thread, this makes sure that if the receiver queue
+          -- thread throws exception while decoding the current thread will
+          -- also die.  For similar reason we are using 'withAsync' to run
+          -- receive queue thread.
+          --
+          -- MonadAsync does not expose 'link' function at the moment
+          -- link queueAsync
+          x <- runPipelinedPeerSender
+            receiveQueue trailingBytes collectQueue
+            codec channel peer
+          cancel queueAsync
+          return x
```

Comment:
This section may already be superseeded by what's in master now. Have a look. We're using `concurrently` from `MonadAsync` so this should all do the right thing. I don't think we need `link` here.

Reply:
The reason for using `withAsync` instead of `concurrently` is that the `runPipelinedPeerReceiverQueue` is just an infinite loop (which might error with a decoder failure).  In tests our sender is often a thread with  a clear termination condition.  So `runPipelinedPeer` will be done when the `runPipelinedPeerSender` is done, while `concurrently` awaits for both threads to terminate.

The link would help when the `runPipelinedPeerReceiverQueue` would fail with a decoder error (which we propagate).  This is done by `concurently` but not by `withAsync`, afaik.

Reply:
Ah yes ok it's not symmetric. So can either make the receiver queue thread shut down too, or can use `withAsync` but I don't think we need link.

Worst case, one uses something like
```
atomically $ do
    ml <- pollSTM left
    mr <- pollSTM right
    case (ml, mr) of
      (Just (Left  e), _) -> throwM e
      (_, Just (Left  e)) -> throwM e
      (_, Just (Right x)) -> return x
      (_,  _)             -> retry

```

Reply:
This snippet would need to run in another thread. We would still need to link it to the main thread, otherwise the main thread will not be killed.

Even if the receiver queue thread shuts down we need to cancel the thread that runs the `PipelinedPeer` (otherwise it will block indefinitely on some STM transaction). If  we don't want `link` (we don't have it :/), we need something like this running in another thread:
```
atomically $ do
    ml <- pollSTM left
    mr <- pollSTM right
    case (ml, mr) of
      (Just (Left  e), _) -> throwTo rightId >> throwM e
      (_, Just (Left  e)) -> throwTo leftId >>  throwM e                              -- 
      (_, Just (Right x)) -> return x
      (_,  _)             -> retry
```
It will kill the other thread, when that happens the main thread will be done and this control thread will be canceled as well (if we use `withAsync).